rinokeras/core/v1x/train/RinokerasGraph.py

```diff
@@ -160,21 +161,22 @@ def run_epoch(self,
                   epoch_num: Optional[int] = None,
                   summary_writer: Optional[tf.summary.FileWriter] = None,
                   save_outputs: Optional[str] = None) -> MetricsAccumulator:
-        all_outputs = []
 
         with self.add_progress_bar(data_len, epoch_num).initialize():
             assert self.epoch_metrics is not None
-            while True:
-                if save_outputs is not None:
-                    loss, outputs = self.run('default', return_outputs=True)
-                    all_outputs.append(outputs)
-                else:
+            if save_outputs is not None:
+                i = 0
+                with h5py.File(save_outputs, 'w') as f:
+                    while True:
+                        loss, outputs = self.run('default', return_outputs=True)
+                        grp = f.create_group(str(i))
+                        outputs = outputs[0]  # can we rely on this being a tuple of length 1?
```

Comment:
@DavidMChan I'm not sure about this, but can I rely on the run output always being a tuple of length 1?

Reply:
Effectively, the outputs are the forward pass of your model. This means that "outputs" can be whatever you want it to be (a numpy array, a dict of numpy arrays, a tuple of arrays, etc.) This is probably why @rmrao hijacked it for use in TAPE. It also makes it tricky to write a generic saving function for the outputs since you have no guarantees on the data format. You can know, however, that the outputs will be the result of a forward run of the model (so they are convertible to tensors). 

Perhaps it makes sense instead of adding options, to add a callback function? Not entirely sure, but this is why we used pickle.

Reply:
I like the idea of a callback function. The `save_outputs` option is really more about quick-and-dirty debugging than it is a real feature at the moment.

Reply:
Otherwise there's not a really good, general way of saving things. It'll vary hugely. Plus a callback would let us do things other than saving them.